File: DynamicProgramming/knapsack/topdownDP.py
#dp O[profits.length][capacity+1s] time O(N*C) space
def solve_knapsack(profits,weights, capacity):
  dp=[[-1 for x in range(capacity+1)] for y in range(len(profits))]
  return knapsack_recursive(dp, profits, weights, capacity, 0)

def knapsack_recursive(dp, profits, weights, capacity, currentIndex):
  #base check
  if capacity <= 0 or currentIndex >= len(profits):
    return 0
  
  #memoize
  if dp[currentIndex][capacity] != -1:
    return dp[currentIndex][capacity]

  profit1=0 #takeit
  if weights[currentIndex] <=capacity:
    profit1 = profits[currentIndex]+ knapsack_recursive(
      dp, profits, weights, capacity-weights[currentIndex], currentIndex+1)

  #leaveit
  profit2=knapsack_recursive(
    dp, profits, weights, capacity, currentIndex+1)

  dp[currentIndex][capacity] = max(profit1,profit2)
  return dp[currentIndex][capacity]


def main():
  print(solve_knapsack([1,6,10,16],[1,2,3,5],7))
  print(solve_knapsack([1,6,10,16],[1,2,3,5],6))


main()

File: DynamicProgramming/knapsack/topdownDP.js
//O(numItems*capacity) time complexity
const solveKnapsack = (profits, weights, capacity) => {
  const dp = [];
  const knapsackRecursive=(profits, weights, capacity, currentIndex) => {
    if(capacity <= 0 || currentIndex >= profits.length) return 0;

    dp[currentIndex] = dp[currentIndex] || [];
    if(typeof dp[currentIndex][capacity] !=='undefined') {
      return dp[currentIndex][capacity];
    }
    let profit1=0;
    if(weights[currentIndex] <= capacity){
      //take it
      profit1 = profits[currentIndex] + knapsackRecursive(profits, weights, capacity - weights[currentIndex], currentIndex+1);
    }
    const profit2= knapsackRecursive(profits, weights, capacity, currentIndex+1);//leave it

    dp[currentIndex][capacity] = Math.max(profit1, profit2);
    return dp[currentIndex][capacity];
  }
  return knapsackRecursive(profits, weights, capacity,0);
};

const profits = [1,6,10,16];
const weights [1,2,3,5];
console.log(`Total knapsack profit: ---> ${solveKnapsack(profits, weights, 7)}`);
console.log(`Total knapsack profit: ---> ${solveKnapsack(profits, weights, 6)}`);
